adapters/repos/db/vector/hnsw/startup.go

```diff
@@ -88,10 +91,34 @@ func (h *hnsw) restoreFromDisk() error {
 			h.metrics.TrackStartupReadCommitlogDiskIO)
 		fdBuf := bufio.NewReaderSize(metered, 256*1024)
 
+		var reader rwhasher.ReaderHasher
+		var checksumReader commitlog.ChecksumReader
+
+		checksumFileName := commitLogChecksumFileName(fileName)
+		hasChecksumFile, err := fileExists(checksumFileName)
+		if err != nil {
+			return errors.Wrapf(err, "verifying checksum file existence %q", checksumFileName)
+		}
+
+		if hasChecksumFile {
+			checksumFile, err := os.Open(checksumFileName)
+			if err != nil {
+				return errors.Wrapf(err, "opening checksum file %q for reading", checksumFileName)
+			}
+
+			defer checksumFile.Close()
+
+			reader = rwhasher.NewCRC32Reader(fdBuf)
+			checksumReader = commitlog.NewCRC32ChecksumReader(bufio.NewReader(checksumFile))
+		} else {
+			reader = commitlog.NewNoopReaderHasher(fdBuf)
+			checksumReader = commitlog.NewNoopChecksumReader()
+		}
+
 		var valid int
-		state, valid, err = NewDeserializer(h.logger).Do(fdBuf, state, false)
+		state, valid, err = NewDeserializer(h.logger).Do(reader, checksumReader, state, false)
 		if err != nil {
-			if errors.Is(err, io.EOF) || errors.Is(err, io.ErrUnexpectedEOF) {
+			if errors.Is(err, io.EOF) || errors.Is(err, io.ErrUnexpectedEOF) || errors.Is(err, commitlog.ErrInvalidChecksum) {
```

Comment:
I think we need to recover differently on ErrInvalidChecksum. You can see below the original recovery is via `os.Truncate` to delete any extra bytes read.

Reply:
the intention was to still use the original truncate up to the point it was successfully read. There the truncate of the checksum file is missing...